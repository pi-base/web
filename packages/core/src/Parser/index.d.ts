declare module 'rehype-truncate' {
  import { Root } from 'mdast'
  import { Transformer } from 'unified'

  type Opts = {
    disable?: boolean
    ellipses?: string
    maxChars: number
    ignoreTags?: string[]
  }

  export default function rehypeTruncate(opts: Opts): Transformer<Root, Root>
}
