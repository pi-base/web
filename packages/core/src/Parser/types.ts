import { TaggedRef } from '../Ref'

type InternalKind = 'S' | 'P' | 'T'
type ExternalKind = TaggedRef['kind']

type Link = { href: string; title: string }
type Linker<T> = (id: T) => Link

type Ids = {
  internal: [InternalKind, string]
  external: [ExternalKind, string]
}

export type Linkers = { [K in keyof Ids]: Linker<Ids[K]> }

export type InternalLinkNode = {
  type: 'internalLink'
  kind: InternalKind
  id: string
}

export type ExternalLinkNode = {
  type: 'externalLink'
  kind: ExternalKind
  id: string
}
