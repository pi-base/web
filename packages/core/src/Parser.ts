import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import rehypeKatex from 'rehype-katex'
import { unified, Processor, Transformer } from 'unified'
import unist from 'unist'
import { visit } from 'unist-util-visit'
import {
  Code,
  CompileContext,
  Effects,
  Extension,
  HtmlExtension,
  Token,
  State,
} from 'micromark-util-types'
import { toH } from 'hast-to-hyperscript'
import h from 'hyperscript'

const dollarSign = 36
const leftBrace = 123
const rightBrace = 125

// TODO: should we support \(...\) and \[...\] style math? Or standardize on $?
const syntax: Extension = {
  text: {
    [leftBrace]: {
      // See https://github.com/micromark/micromark#extending-markdown
      //
      // Aims to parse
      //    {value}  => { type: 'internalLink', value }
      //   {{value}} => { type: 'citation', value }
      tokenize(effects: Effects, ok: State, nok: State) {
        let external = false

        // HACK
        // For some reason, fromMarkdown's this.buffer() / this.resume() does
        // not appear to actually buffer and return the contents we're consuming.
        //
        // For now we'll manually track them while scanning, and emit a
        // synthetic event with the full context _after_ the scan.
        const buffer: number[] = []

        function start(code: Code) {
          if (code !== leftBrace) {
            return nok(code) // Expected {
          }

          effects.enter('citation')
          effects.enter('citationMarker')
          effects.consume(code)
          return nesting
        }

        function nesting(code: Code) {
          if (code === leftBrace) {
            if (external) {
              return nok(code) // Unexpected {{{
            }

            external = true
            effects.consume(code)

            effects.exit('citationMarker')
            // HACK: we're manually buffering, so can't emit this event yet
            // effects.enter('citationString', { type: 'citation' })
            return body
          } else {
            effects.exit('citationMarker')
            // HACK: we're manually buffering, so can't emit this event yet
            // effects.enter('citationString', { type: 'internalLink' })
            return body(code)
          }
        }

        function body(code: Code) {
          if (code === null) {
            return nok(code) // Unexpected EOF
          }

          if (code === rightBrace) {
            // HACK: per above, we can now manually emit the buffered content
            const value = String.fromCharCode(...buffer)
            effects.enter('citationString', {
              external,
              value,
            })
            effects.exit('citationString')
            effects.enter('citationMarker')
            return close(code)
          }

          buffer.push(code)
          effects.consume(code)
          return body
        }

        function close(code: Code) {
          if (code !== rightBrace) {
            return nok(code)
          }

          if (external) {
            external = false
            effects.consume(code)
            return close
          }

          effects.consume(code)
          effects.exit('citationMarker')
          effects.exit('citation')
          return ok
        }

        return start
      },
    },
    [dollarSign]: {
      tokenize: tokenizeMath,
    },
  },
}

function tokenizeMath(effects: Effects, ok: State, nok: State) {
  let display = false
  const buffer: number[] = []

  function start(code: Code) {
    if (code !== dollarSign) {
      return nok(code)
    }

    effects.enter('math')
    effects.enter('mathMarker')
    effects.consume(code)
    return nesting
  }

  function nesting(code: Code) {
    if (code === dollarSign) {
      if (display) {
        return nok(code) // Unexpected $$$
      }

      display = true
      effects.consume(code)
    }

    effects.exit('mathMarker')
    return body
  }

  function body(code: Code) {
    if (code === null) {
      return nok(code) // Unexpected EOF
    }

    if (code === dollarSign) {
      const value = String.fromCharCode(...buffer)
      effects.enter('mathString', {
        display,
        value,
      })
      effects.exit('mathString')
      effects.enter('mathMarker')
      return close(code)
    }

    buffer.push(code)
    effects.consume(code)
    return body
  }

  function close(code: Code) {
    if (code !== dollarSign) {
      return nok(code)
    }

    if (display) {
      display = false
      effects.consume(code)
      return close
    }

    effects.consume(code)
    effects.exit('citationMarker')
    effects.exit('citation')
    return ok
  }

  return start
}

const fromMarkdown: HtmlExtension = {
  enter: {
    citationString(this: CompileContext, token: Token) {
      // HACK: a few things about the types don't seem to actually track with
      // the runtime behavior
      //
      // - buffer / resume does not appear to work, per the above (although this
      //   may be a configuration issue)
      // - we don't _want_ to use this.raw, as we're not emitting a raw string;
      //   rather, we want to pass { type, value } tokens on for downstream
      //   consumers
      // - token will have the extra fields specified during the `enter` that
      //   emitted them
      //
      /* eslint-disable */
      // @ts-ignore
      const { external, value } = token
      // @ts-ignore
      this.enter({ type: external ? 'citation' : 'internalLink', value }, token)
      /* eslint-disable */
      this.buffer()
    },
    mathString(this: CompileContext, token: Token) {
      /* eslint-disable */
      // @ts-ignore
      const { display, value } = token
      // @ts-ignore
      this.enter({ type: display ? 'blockMath' : 'inlineMath', value }, token)
      /* eslint-disable */
      this.buffer()
    },
  },
  exit: {
    citationString(this: CompileContext, token: Token) {
      this.resume()
      // @ts-ignore
      this.exit(token)
    },
    mathString(this: CompileContext, token: Token) {
      this.resume()
      // @ts-ignore
      this.exit(token)
    },
  },
}

export function tokenize(this: Processor): void {
  const data = this.data()

  add('micromarkExtensions', syntax)
  add('fromMarkdownExtensions', fromMarkdown)

  function add(field: string, value: unknown) {
    const found = data[field]

    if (found && found instanceof Array) {
      found.push(value)
    } else {
      data[field] = [value]
    }
  }
}

export function transformTokens(this: Processor): Transformer {
  return function transformer(tree: unist.Node) {
    visit<unist.Node & { value?: unknown }, string>(
      tree,
      'citation',
      (node) => {
        node.data = {
          hName: 'citation',
          hChildren: [{ type: 'text', value: node.value }],
        }
      },
    )

    visit<unist.Node & { value?: unknown }, string>(
      tree,
      'internalLink',
      (node) => {
        node.data = {
          hName: 'internalLink',
          hChildren: [{ type: 'text', value: node.value }],
        }
      },
    )

    visit<unist.Node & { value?: unknown }, string>(
      tree,
      'blockMath',
      (node) => {
        node.data = {
          hName: 'span',
          hProperties: {
            className: ['math-display'],
          },
          hChildren: [{ type: 'text', value: node.value }],
        }
      },
    )

    visit<unist.Node & { value?: unknown }, string>(
      tree,
      'inlineMath',
      (node) => {
        node.data = {
          hName: 'span',
          hProperties: {
            className: ['math-inline'],
          },
          hChildren: [{ type: 'text', value: node.value }],
        }
      },
    )
  }
}

export async function renderHTML(processor: Processor, body: string) {
  const parsed = await processor.run(processor.parse(body))
  // @ts-ignore
  return toH(h, parsed).innerHTML
}

export function tokenized(): Processor {
  return unified().use(remarkParse).use(tokenize)
}

export function html(): Processor {
  return (
    tokenized()
      .use(transformTokens)
      // @ts-ignore
      .use(remarkRehype)
      // @ts-ignore
      .use(rehypeKatex)
  )
}
