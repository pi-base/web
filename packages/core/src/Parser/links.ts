import type {
  CompileContext,
  Extension as FromMarkdownExtension,
} from 'mdast-util-from-markdown'
import type { Code, Construct, Token, Tokenizer } from 'micromark-util-types'
import type { Plugin, Processor } from 'unified'
import createDebug from 'debug'

import { ExternalLinkNode, InternalLinkNode } from './types'

const debug = createDebug('pi-base:links')

/**
 * Extends the Markdown parser, adding support for tokenizing {internal} and
 * {{external}} links into InternalLinkNode and ExternalLinkNode tokens.
 *
 * See https://github.com/micromark/micromark#creating-a-micromark-extension for
 * details on the micromark extension API.
 */
const cr = -5
const lf = -4
const crlf = -3
const eof = null
const colon = 58
const leftBrace = 123
const rightBrace = 125

function prefixKind(code: Code) {
  switch (code) {
    case 80: // P
    case 112: // p
      return 'P'
    case 83: // S
    case 115: // s
      return 'S'
    case 84: // T
    case 116: // t
      return 'T'
    default:
      return null
  }
}

const tokenize: Tokenizer = (effects, ok, nok) => {
  debug('start tokenize')
  /* The tokenizer is a state machine that starts in the `start` state,
   * accumulating text and transitioning as described above the other states
   * below.
   */
  return start

  /* Consumes the single starting { */
  function start(code: Code) {
    debug('start %s', code)
    effects.enter('linkMarker')
    effects.consume(code)

    return internalOrExternal
  }

  /* Consumes a single character
   * - if {, parse the rest as an external link
   * - otherwise parse an interal link with kind determined by the consumed
   *   character [PST]
   */
  function internalOrExternal(code: Code) {
    debug('intenalOrExternal %s', code)
    if (code === leftBrace) {
      effects.consume(code)
      effects.exit('linkMarker')

      effects.enter('externalLink')
      effects.enter('externalKind')
      effects.enter('chunkString', { contentType: 'string' })
      return externalKind
    }

    effects.exit('linkMarker')
    effects.enter('internalKind')
    effects.consume(code)
    effects.exit('internalKind')
    const kind = prefixKind(code)
    if (kind) {
      effects.enter('internalLink', { kind })
      effects.enter('chunkString', { contentType: 'string' })
      return internalId
    }

    return nok(code)
  }

  /* Consume word characters up to a :, then transition to externalId */
  function externalKind(code: Code) {
    debug('externalKind %s', code)
    if (code === cr || code === lf || code === crlf || code === eof) {
      return nok(code)
    }

    if (code === colon) {
      effects.exit('chunkString')
      effects.exit('externalKind')
      effects.enter('externalSeparator')
      effects.consume(code)
      effects.exit('externalSeparator')
      effects.enter('externalId')
      effects.enter('chunkString', { contentType: 'string' })
      return externalId
    }

    effects.consume(code)
    return externalKind
  }

  /* Consume word characters up to a }, then finalize */
  function internalId(code: Code) {
    debug('internalId %s', code)
    if (code === cr || code === lf || code === crlf || code === eof) {
      return nok(code)
    }

    if (code === rightBrace) {
      effects.exit('chunkString')
      effects.exit('internalLink')
      effects.enter('linkMarker')
      return close(code)
    }

    effects.consume(code)
    return internalId
  }

  /* Consume word characters up to a }, then finalize */
  function externalId(code: Code) {
    debug('externalId %s', code)
    if (code === cr || code === lf || code === crlf || code === eof) {
      return nok(code)
    }

    if (code === rightBrace) {
      effects.exit('chunkString')
      effects.exit('externalId')
      effects.exit('externalLink')
      effects.enter('linkMarker')
      return close(code)
    }

    effects.consume(code)
    return externalId
  }

  /* Consume any leftover }s */
  function close(code: Code) {
    debug('close %s', code)
    if (code === rightBrace) {
      effects.consume(code)
      return close
    }
    effects.exit('linkMarker')
    return ok(code)
  }
}

/**
 * Helpers for compiling emitted tokens into AST nodes
 *
 * The `as` castings here aren't ideal, but the MDAST types are very rigid and
 * expect a closed enum of Node types. The underlying implementation appears to
 * support `enter`ing other shapes of nodes, but we should be careful about this
 * breaking in dependency updates.
 *
 * TODO: improve error handling throughout here (i.e. don't silently ignore
 * typechecking errors)
 */
function enterExternalLink(this: CompileContext, token: Token) {
  this.enter<any>(
    {
      type: 'externalLink',
    } as Omit<ExternalLinkNode, 'id' | 'kind'>,
    token,
  )
  this.buffer()
}

function enterExternalKind(this: CompileContext, _token: Token) {
  this.buffer()
}

function exitExternalKind(this: CompileContext, _token: Token) {
  const kind = this.resume()
  const node: unknown = this.stack[this.stack.length - 2]
  if (isExternalLink(node) && isExternalKind(kind)) {
    node.kind = kind
  }
}

function enterExternalId(this: CompileContext, _token: Token) {
  this.buffer()
}

function exitExternalId(this: CompileContext, _token: Token) {
  const id = this.resume()
  const node: unknown = this.stack[this.stack.length - 2]
  if (isExternalLink(node)) {
    node.id = id
  }
}

function exitExternalLink(this: CompileContext, token: Token) {
  this.resume()
  this.exit(token)
}

function enterInternalLink(this: CompileContext, token: Token) {
  this.enter<any>(
    {
      type: 'internalLink',
      kind: (token as any).kind,
    } as Omit<InternalLinkNode, 'id'>,
    token,
  )
  this.buffer()
}

function exitInternalLink(this: CompileContext, token: Token) {
  const id = this.resume()
  const node: unknown = this.exit(token)
  if (isInternalLink(node)) {
    node.id = id
  }
}

/**
 * Register the plugin with micromark and unist
 */
const construct: Construct = {
  name: 'links',
  tokenize,
}

const syntax = {
  text: {
    [leftBrace]: construct,
  },
}

const fromMarkdown: FromMarkdownExtension = {
  enter: {
    externalKind: enterExternalKind,
    externalId: enterExternalId,
    externalLink: enterExternalLink,
    internalLink: enterInternalLink,
  },
  exit: {
    externalKind: exitExternalKind,
    externalId: exitExternalId,
    externalLink: exitExternalLink,
    internalLink: exitInternalLink,
  },
}

function isExternalLink(node: any): node is ExternalLinkNode {
  return node && node.type === 'externalLink'
}

function isExternalKind(kind: any): kind is ExternalLinkNode['kind'] {
  return kind && ['doi', 'wikipedia', 'mr', 'mathse', 'mo', 'zb'].includes(kind)
}

function isInternalLink(node: any): node is InternalLinkNode {
  return node && node.type === 'internalLink'
}

export const links: Plugin = function (this: Processor) {
  const data = this.data()

  add('micromarkExtensions', syntax)
  add('fromMarkdownExtensions', fromMarkdown)

  function add(field: string, value: unknown) {
    const list = data[field] ? data[field] : (data[field] = [])

    ;(list as unknown[]).push(value)
  }
}
