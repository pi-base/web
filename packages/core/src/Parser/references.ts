import { Transformer } from 'unified'
import { Root } from 'hast'
import { visit } from 'unist-util-visit'
import { ExternalLinkNode, InternalLinkNode, Linkers } from './types'

/**
 * Builds an MDAST transformer that enriches parsed internal and external link
 * nodes used provided link builders, and adds metadata to the MDAST so that
 * they will convert correctly to HAST.
 *
 * See https://github.com/syntax-tree/mdast-util-to-hast#fields-on-nodes
 */
export function references({ internal, external }: Linkers) {
  return (): Transformer<Root, Root> => {
    return function transformer(tree: Root) {
      visit(tree, 'internalLink', (node: InternalLinkNode) => {
        const { kind, id } = node

        const { href, title } = internal([kind, id])

        Object.assign(node, {
          data: {
            hName: 'a',
            hProperties: {
              href,
              title,
              className: 'internal-link',
            },
            hChildren: [
              {
                type: 'text',
                value: title,
              },
            ],
          },
        })
      })

      visit(tree, 'externalLink', (node: ExternalLinkNode) => {
        const { href, title } = external([node.kind, node.id])

        Object.assign(node, {
          data: {
            hName: 'a',
            hProperties: {
              href,
              title,
              className: 'external-link',
            },
            hChildren: [
              {
                type: 'text',
                value: title,
              },
            ],
          },
        })
      })
    }
  }
}
