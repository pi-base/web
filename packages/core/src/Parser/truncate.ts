import { Root } from 'mdast'
import { Transformer } from 'unified'

type Opts = {
  disable?: boolean
  ellipses?: string
  maxChars?: number
  ignoreTags?: string[]
}

// Adapted from https://github.com/luk707/rehype-truncate
export function truncate({
  disable = false,
  ellipses = '\u2026',
  ignoreTags = [],
  maxChars = 120,
}: Opts = {}): Transformer<Root, Root> {
  return truncator

  function truncator(tree: Root) {
    if (!disable) {
      truncateNode(tree)
    }
  }

  function truncateNode(node: any, tf = 0) {
    let foundText = tf

    if (node.type === 'text') {
      foundText += node.value.length

      if (foundText >= maxChars) {
        node.value = `${node.value.slice(
          0,
          node.value.length - (foundText - maxChars),
        )}${ellipses}`
        return maxChars
      } else if (node.value === '\n') {
        node.value = ''
        return maxChars
      }
    }

    if (node.type === 'root' || node.type === 'element') {
      if (node.type === 'element' && ignoreTags.includes(node.tagName)) {
        return foundText
      }
      for (let i = 0; i < node.children.length; i++) {
        if (foundText === maxChars) {
          node.children.splice(i, 1)
          i--
          continue
        }
        foundText = truncateNode(node.children[i], foundText)
      }
    }

    return foundText
  }
}
