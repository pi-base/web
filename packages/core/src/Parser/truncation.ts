import { Root, Content } from 'mdast'
import { Transformer } from 'unified'

/**
 * Collapses long descriptions down to a reasonable preview, adding an ellipsis.
 */
function gather(nodes: Content[], to: number) {
  let length = 0
  const acc = []

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]

    if (node.type === 'text') {
      const value = node.value || ''
      if (value.length + length >= to) {
        const fragment = value.slice(0, to - length - value.length) + '...'
        acc.push({ ...node, value: fragment })
        return acc
      } else {
        acc.push(node)
        length = length + value.length + 1
      }
    } else {
      if (length + 3 >= to) {
        acc.push(node)
        acc.push({ type: 'text', value: '...' })
        return acc
      } else {
        acc.push(node)
        length = length + 3
      }
    }
  }

  return acc
}

export function truncation(to = 100): Transformer<Root, Root> {
  return function transformer(tree: Root) {
    const node = tree.children[0] || {}

    if ('children' in node) {
      return {
        ...node,
        type: 'root',
        children: gather(node.children, to),
      } as Root
    } else {
      return { ...node, type: 'root' } as Root
    }
  }
}
