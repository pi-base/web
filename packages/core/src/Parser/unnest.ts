import { Node } from 'unist'
import { Transformer } from 'unified'

type Tree = Node & { children?: Tree[] }

/**
 * Many default parsers create an AST that looks like
 * 
 *     root:
 *       children:
 *         - type: paragraph
 *           children:
 *             - type: singleton
 * 
 * which renders as <p><singleton/></p>. This transformer removes the wrapping
 * <p/> node 
 * 
 *     root:
 *       children:
 *         - type: singleton
 * 
 * to allow for inline-only <singleton/> elements.
 */
export const unnest = () => {
  const transformer: Transformer<Tree> = (tree) => {
    if (
      tree &&
      tree.children?.length === 1 &&
      tree.children[0].type === 'paragraph' &&
      tree.children[0].children?.length === 1
    ) {
      return {
        ...tree,
        children: tree.children[0].children,
      }
    }

    return tree
  }

  return transformer
}
