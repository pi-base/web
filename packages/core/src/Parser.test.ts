import { Node } from 'unist'
import { visit } from 'unist-util-visit'
import { tokenized, html, renderHTML } from './Parser'

function deposition(tree: Node) {
  visit(tree, (node) => delete node['position'])
  return tree
}

describe('tokenization', () => {
  function run(input: string) {
    const processor = tokenized()
    return deposition(processor.runSync(processor.parse(input)))
  }

  it('parses a citation', () => {
    expect(run('{{doi:123}}')).toEqual({
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'citation',
              value: 'doi:123',
            },
          ],
        },
      ],
    })
  })

  it('parses an internal link', () => {
    expect(run('{S1}')).toEqual({
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'internalLink',
              value: 'S1',
            },
          ],
        },
      ],
    })
  })

  it('parses math', () => {
    expect(run('$1 + 1 = 2$ and $$2 + 2 = 4$$')).toEqual({
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'inlineMath',
              value: '1 + 1 = 2',
            },
            {
              type: 'text',
              value: ' and ',
            },
            {
              type: 'blockMath',
              value: '2 + 2 = 4',
            },
          ],
        },
      ],
    })
  })
})

describe('html', () => {
  async function run(body: string) {
    return renderHTML(html(), body)
  }

  it('renders reference links', async () => {
    expect(await run('{S1} is described in {{mr:123}}')).toEqual(
      '<internalLink>S1</internalLink> is described in <citation>mr:123</citation>',
    )
  })

  it('renders a complex example', async () => {
    const example = `This is a list of links

* {S000123}
* {P000123}
* {T000123}
* {{doi:123}}
* {{mr:123}}
* {{wikipedia:123}}
* {{mathse:123}}
* {{mo:123}}`

    expect(await run(example)).toEqual(`<p>This is a list of links</p>
<ul>
<li><internalLink>S000123</internalLink></li>
<li><internalLink>P000123</internalLink></li>
<li><internalLink>T000123</internalLink></li>
<li><citation>doi:123</citation></li>
<li><citation>mr:123</citation></li>
<li><citation>wikipedia:123</citation></li>
<li><citation>mathse:123</citation></li>
<li><citation>mo:123</citation></li>
</ul>`)
  })

  it('renders a complex example with math', async () => {
    const example = `Inline math $2 + 2 = 4$ and display math $$2 + 2 = 4$$.

This is a list of links

* {S000123}
* {P000123}
* {T000123}
* {{doi:123}}
* {{mr:123}}
* {{wikipedia:123}}
* {{mathse:123}}
* {{mo:123}}`

    expect(await run(example)).toMatchSnapshot()
  })
})
