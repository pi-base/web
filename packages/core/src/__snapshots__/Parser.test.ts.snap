// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html renders a complex example with math 1`] = `
"<p>Inline math <span class=\\"math-inline\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.72777em;vertical-align:-0.08333em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">4</span></span></span></span></span> and display math <span class=\\"math-display\\"><span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.72777em;vertical-align:-0.08333em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222222222222222em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;:;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.64444em;vertical-align:0em;:;\\"></span><span class=\\"mord\\">4</span></span></span></span></span></span>.</p>
<p>This is a list of links</p>
<ul>
<li><internalLink>S000123</internalLink></li>
<li><internalLink>P000123</internalLink></li>
<li><internalLink>T000123</internalLink></li>
<li><citation>doi:123</citation></li>
<li><citation>mr:123</citation></li>
<li><citation>wikipedia:123</citation></li>
<li><citation>mathse:123</citation></li>
<li><citation>mo:123</citation></li>
</ul>"
`;
